### PR TITLE
MORE text output engine fixes

### PR DESCRIPTION
**Pull request by FeralChild64**
Improve MORE command/engine output:

- fix for top line cut if help string is exactly 24 lines long (bug noticed by @Kappa971), also fixed the situation when last (23rd) line of input is exactly 80 characters long
- command help now detects whether output is redirected and acts accordingly (requested by @Kappa971)
- improved the output a little, got rid of unnecessary empty lines at output start/end if `command | MORE` is executed

### CODE REVIEW

**Comment by kcgen** [comment]
Very nice handling when re-directing color codes to file (which is currently a problem) :+1:

**Comment by FeralChild64** [comment]
@kcgen Unfortunately, it only tracks ANSI sequences for purpose of detecting output redirection. Stripping the ANSI sequences from redirected output would require much more work, I'm not even sure if it could be done reliably.
1st problem is that we would need to strip them from translatable strings, 2nd - we need some text to be sent out to detect redirection, if the stream starts with ANSI sequence we won't know whether we should strip it away or not.

BTW, MORE from MS DOS 6.22 is unable to detect output redirection at all, so we are better :)

File: src/dos/program_more_output.cpp [hunk 1]
```diff
@@ -133,31 +158,56 @@ UserDecision MoreOutputBase::DisplaySingleStream()
...
 		}
 		if (code != '\n') {
 			was_prompt_recently = false;
 		}
-		if (is_last) {
+		if (is_last && !line_overflow) {
 			// Skip further processing (including possible user prompt)
-			// if we know no data is left
+			// if we know no data is left and no line overflow produced
+			// an 'artificial new line'
```
Comment (johnnovak) [inline src/dos/program_more_output.cpp:210]: Seems fine at the surface level, but I find it hard to really understand how the various flags control the logic, whether the order of `if/else if` statement evaluations matter, etc. Given that you're checking the state of multiple flags in quite a few conditional statements, I'm suspecting what you really have here is a bunch of states that you're encoding with different flag combinations.

Wouldn't making the state explicit in the form of a state-machine make the code easier to undestand and maintain? As opposed to having implicit states in the form of various "undocumented" flag-combinations. Could you introduce a state enum, then convert this whole thing into an FSM?

File: src/dos/program_more_output.h [hunk 2]
```diff
@@ -39,6 +39,8 @@ class MoreOutputBase {
 	virtual void Display() = 0;
 
 protected:
+	static uint8_t GetCurrentColumn();
+	static uint8_t GetCurrentRow();
```
Comment (johnnovak) [inline src/dos/program_more_output.h:43]: Hmm, so are these zero-indexed? (I assume yes)
Is "current" the position where the last character was written to, or where the next character would land at? If the latter, I'm guessing the current row gets auto-incremented when the end of the line has been reached?
I think some explanation is in order for these for the benefit of future maintainers.

File: src/dos/program_more_output.cpp (hunk 1 above)
Comment (kcgen) [reply src/dos/program_more_output.cpp:210]: Definitely not the first time the CLI and shell parsing code has gotten quite tricky.

Fortunately this is readable and safe (unlike previously, where we had states plus lots of C char* pointer arithmetic), none the less, it's still juggling a lot of bools and branches.

@FeralChild64 - is it possible to feed content to `program_more_*` without standing up a whole dosbox instance? It might be a good candidate for unit tests.

The unit tests would let you hit all those state permutations (and variations on them too, illegal content, etc).

File: src/dos/program_more_output.h (hunk 2 above)
Comment (FeralChild64) [reply src/dos/program_more_output.h]: Changed names to `GetCursorColumn` and `GetCursorRow`. Hopefully it is more clear now what these functions are doing.

File: src/dos/program_more_output.cpp [hunk 3]
```diff
@@ -183,17 +234,38 @@ UserDecision MoreOutputBase::DisplaySingleStream()
...
+	if (column_start == column_end) {
+		// Usually redirected output should be detected
+		// till this point, but in a VERY special case
+		// (only carriage return and ANSI sequences in
+		// the input till now, cursor in one of the two
+		// last rows, no file/device as a MORE command
+		// argument) it will only be detected here
+		WriteOut("\n");
+		is_output_redirected = true;
+		return UserDecision::Continue;
```
[the remark this answers is not in the document]
Comment (FeralChild64) [reply src/dos/program_more_output.cpp:257]: The redirected output flag is basically independent of all the other states - I don;t see any possibility to introduce a state for it.

File: src/dos/program_more_output.cpp (hunk 1 above)
Comment (FeralChild64) [reply src/dos/program_more_output.cpp:200]: @johnnovak: I have changed the code to use a state variable - but this was a huge and risky change, it will take quite some time to test it.

@kcgen: The MORE text engine tracks the cursor position after printing out characters - it would be hard to test it using unit tests, one would need to have a good simulation of the whole screen output, including ANSI handling.

**Comment by FeralChild64** [comment]
> > ```
> > * fix for top line cut if help string is exactly 23 lines long
> > ```
> 
> *24 lines

OK, fixed.

File: src/dos/program_more_output.h (hunk 2 above)
Comment (johnnovak) [reply src/dos/program_more_output.h]: > Changed names to `GetCursorColumn` and `GetCursorRow`. Hopefully it is more clear now what these functions are doing.

Yes, definitely it's clear now!

File: src/dos/program_more_output.cpp (hunk 1 above)
Comment (johnnovak) [reply src/dos/program_more_output.cpp:200]: > I have changed the code to use a state variable - but this was a huge and risky change, it will take quite some time to test it.

Great, I think it reads a lot better, and it's much easier to reason about the transitions of a single state variable than  trying to visualise all possible permutations of 5+ boolean flags in your head. Even if it requires a bit more testing effort short term, the long-term maintenance aspects and the increased code clarity (which reduces the chance of weird bugs) makes it well worth it.

Thanks for doing this change @FeralChild64!

**Review by johnnovak — APPROVED** [review]
Code looks great @FeralChild64, approving this 👍🏻 I don't have time to test this now, so feel free to merge once you think it's solid enough, then we can address any potential edge-case later.

File: src/dos/program_more_output.cpp (hunk 1 above)
Comment (kcgen) [reply src/dos/program_more_output.cpp:200]: > The MORE text engine tracks the cursor position after printing out characters - it would be hard to test it using unit tests, one would need to have a good simulation of the whole screen output, including ANSI handling.

Thanks for explaining that - and yes, that does sound excessively involved for testing (and at the same time, helps me understand this better too).

**Review by kcgen — APPROVED** [review]
Update looks good! Thanks @FeralChild64.

**Comment by FeralChild64** [comment]
@kcgen But please don’t merge it yet - it is likely there will be some minor changes.

**Comment by kcgen** [comment]
> it is likely there will be some minor changes.

Roger that; will let you merge when ready.

**Comment by FeralChild64** [comment]
@kcgen Did some minor corrections. From my point of view, it is now ready to be merged.
